File: Data Structures & Algorithms/remove-node-from-end-of-linked-list/submission-4.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */

class Solution {
public:
    ListNode* removeNthFromEnd(ListNode* head, int n) {
        auto dummy = new ListNode(0, head);
        auto fast = dummy;
        for (int i = 0; i <= n; ++i) {
            fast = fast->next;
        }
        auto slow = dummy;
        while (fast) {
            slow = slow->next;
            fast = fast->next;
        }
        auto toDelete = slow->next;
        slow->next = toDelete->next;
        delete toDelete;
        auto toReturn = dummy->next;
        delete dummy;
        return toReturn;
    }
};

File: Data Structures & Algorithms/reorder-linked-list/submission-4.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next

class Solution:
    def reverseList(self, head: Optional[ListNode]) -> Optional[ListNode]:
        prev, curr = None, head
        while curr:
            next, curr.next = curr.next, prev
            prev, curr = curr, next
        return prev

    def reorderList(self, head: Optional[ListNode]) -> None:
        slow, fast = head, head.next
        while fast and fast.next:
            slow, fast = slow.next, fast.next.next

        list1, list2 = head, self.reverseList(slow.next)
        slow.next = None

        while list1 and list2:
            next1, next2 = list1.next, list2.next
            list1.next = list2
            list2.next = next1 if next1 else next2
            list1, list2 = next1, next2